### PR TITLE
tests: moving journalctl.sh to a new journal-state tool

### PR DESCRIPTION
**Pull request by sergiocazzolato**
The idea is to use the journalctl tool to query and check the journal logs. 

Also it is included a new test to check the journal-state tool works properly.

### CODE REVIEW

File: tests/lib/tools/suite/journal-state/task.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+summary: smoke test for the journal-state tool
+execute: |
+    # Check help
```
Comment (zyga) [inline tests/lib/tools/suite/journal-state/task.yaml:3]: Could we also shellcheck the program here, even if it is also checked by some other layer.

File: tests/lib/tools/journal-state [hunk 2]
```diff
@@ -73,7 +79,29 @@ get_journalctl_log_from_cursor(){
...
 
-sync_journalctl_log(){
+_sync_journalctl_log(){
     journalctl --flush || true
     journalctl --sync || true
 }
+
+if [ $# -eq 0 ]; then
+    show_help
+    exit 1
```
Comment (zyga) [inline tests/lib/tools/journal-state:89]: Nitpick, usually tools don't fail in this case, only in the case when the arguments are actually invalid.

File: tests/lib/tools/journal-state [hunk 3]
```diff
@@ -73,7 +79,29 @@ get_journalctl_log_from_cursor(){
...
+            ;;
+        *)
+            action=$(echo "$1" | tr '-' '_')
+            shift
+            break
+            ;;
+    esac
+done
+
+"$action" "$@"
```
Comment (zyga) [inline tests/lib/tools/journal-state:107]: How does this behave on unknown action?

File: tests/lib/tools/journal-state [hunk 4]
```diff
@@ -1,17 +1,23 @@
...
 
-# shellcheck source=tests/lib/state.sh
-. "$TESTSLIB/state.sh"
-
 # This file contains all the cursors used for the different tests
-JOURNALCTL_CURSOR_FILE="$RUNTIME_STATE_PATH"/journalctl_cursor
+JOURNALCTL_CURSOR_FILE="$SPREAD_PATH"/tests/runtime-state/journalctl_cursor
+
+show_help() {
+    echo "usage: journal-state start-new-journalctl-log"
```
[the remark this answers is not in the document]
Comment (sergiocazzolato) [reply tests/lib/tools/journal-state:7]: done

File: tests/lib/tools/journal-state [hunk 5]
```diff
@@ -73,7 +79,29 @@ get_journalctl_log_from_cursor(){
...
+
+if [ $# -eq 0 ]; then
+    show_help
+    exit 1
+fi
+
+action=
+while [ $# -gt 0 ]; do
+    case "$1" in
+        -h|--help)
```
[the remark this answers is not in the document]
Comment (sergiocazzolato) [reply tests/lib/tools/journal-state:95]: done

File: tests/lib/tools/suite/journal-state/task.yaml [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    "$TESTSTOOLS"/journal-state -h | MATCH "usage: journal-state start-new-journalctl-log"
```
[the remark this answers is not in the document]
Comment (sergiocazzolato) [reply tests/lib/tools/suite/journal-state/task.yaml:4]: done

File: tests/lib/tools/suite/journal-state/task.yaml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    "$TESTSTOOLS"/journal-state --help | MATCH "usage: journal-state start-new-journalctl-log"
+
+    cursor1=$("$TESTSTOOLS"/journal-state get-last-journalctl-cursor)
+    test $(journalctl --cursor "$cursor1" | grep -c "$SPREAD_JOB") -eq 0
```
[the remark this answers is not in the document]
Comment (sergiocazzolato) [reply tests/lib/tools/suite/journal-state/task.yaml]: As here we are doing grep after with the spread_job it shouldn't fail, the only way it could fail is we run in parallel in the same machine/instance this test but spread does not allow that. 
I added a comment now for that line

File: tests/lib/tools/suite/journal-state/task.yaml [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+    # Check the test is correctly started in the journal log
+    "$TESTSTOOLS"/journal-state start-new-journalctl-log
+    retry --wait 1 --attempts 30 sh -c "test $(journalctl --cursor $cursor1 | grep -c $SPREAD_JOB) -eq 1"
+
+    # check the subcommand check-journalctl-ready works well
+    "$TESTSTOOLS"/journal-state check-journalctl-ready
```
[the remark this answers is not in the document]
Comment (sergiocazzolato) [reply tests/lib/tools/suite/journal-state/task.yaml:15]: done

File: tests/lib/tools/journal-state [hunk 9]
```diff
@@ -1,17 +1,23 @@
...
-
 # This file contains all the cursors used for the different tests
-JOURNALCTL_CURSOR_FILE="$RUNTIME_STATE_PATH"/journalctl_cursor
+JOURNALCTL_CURSOR_FILE="$SPREAD_PATH"/tests/runtime-state/journalctl_cursor
+
+show_help() {
+    echo "usage: journal-state start-new-journalctl-log"
+    echo "usage: journal-state check-journalctl-ready"
+    echo "usage: journal-state get-last-journalctl-cursor"
+    echo "usage: journal-state check-journalctl-log [expression]"
```
[the remark this answers is not in the document]
Comment (sergiocazzolato) [reply tests/lib/tools/journal-state:10]: I just modified the name. Now it is match-log, does it make sense?

File: tests/lib/tools/suite/journal-state/task.yaml (hunk 1 above)
Comment (sergiocazzolato) [reply tests/lib/tools/suite/journal-state/task.yaml:3]: done

File: tests/lib/tools/journal-state (hunk 2 above)
Comment (sergiocazzolato) [reply tests/lib/tools/journal-state:89]: done

File: tests/lib/tools/journal-state (hunk 3 above)
Comment (sergiocazzolato) [reply tests/lib/tools/journal-state:107]: Added a check to validate the command is supported.

**Comment by sergiocazzolato** [comment]
@zyga thanks for the review. The pr has been updated.

File: tests/lib/tools/journal-state (hunk 9 above)
[the remark this answers is not in the document]
Comment (zyga) [reply tests/lib/tools/journal-state]: Yeah, that's nicer. Thanks

File: spread.yaml [hunk 10]
```diff
@@ -499,7 +497,7 @@ debug-each: |
         fi
 
         echo '# journal messages for snapd'
-        get_journalctl_log -u snapd
+        "$TESTSTOOLS"/journal-state get-journalctl-log -u snapd
```
Comment (zyga) [inline spread.yaml:500]: Leftover old name? This should be `get-log`, right?

File: tests/lib/tools/journal-state [hunk 11]
```diff
@@ -1,19 +1,25 @@
...
-# shellcheck source=tests/lib/state.sh
-. "$TESTSLIB/state.sh"
-
 # This file contains all the cursors used for the different tests
-JOURNALCTL_CURSOR_FILE="$RUNTIME_STATE_PATH"/journalctl_cursor
+JOURNALCTL_CURSOR_FILE="$SPREAD_PATH"/tests/runtime-state/journalctl_cursor
+
+show_help() {
+    echo "usage: journal-state start-new-log"
+    echo "usage: journal-state check-ready"
```
Comment (zyga) [inline tests/lib/tools/journal-state:8]: Out of the whole set the only name that doesn't immediately make sense to me is `check-ready`. Perhaps `is-working` or something similar. This function is only called in one place so it can have a longer name.

Comment (sergiocazzolato) [reply tests/lib/tools/journal-state:8]: Already moved to check-log-started, thanks for the review

File: spread.yaml (hunk 10 above)
Comment (sergiocazzolato) [reply spread.yaml:500]: yes, forgot to update that one, thanks!

File: tests/core/enable-disable-units-gpio/task.yaml [hunk 12]
```diff
@@ -57,13 +57,10 @@ execute: |
...
     fi
 
-    # shellcheck source=tests/lib/journalctl.sh
-    . "$TESTSLIB/journalctl.sh"
-
     echo "Then the snap service units concerning the gpio device must be run before and after a reboot"
     expected="Unit snap.core.interface.gpio-100.service has finished starting up"
-    for i in $(seq 120); do
-        get_journalctl_log -xe --no-pager | MATCH "$expected"
+    for _ in $(seq 120); do
```
Comment (anonymouse64) [inline tests/core/enable-disable-units-gpio/task.yaml:62]: doesn't shellcheck complain about unused variables in this instance? 

also not for this test, but would be nice to adapt all such instances of `for ... in $(seq XYZ); do ...; sleep ...; done` to use `retry` instead

File: tests/main/snap-userd-desktop-app-autostart/task.yaml [hunk 13]
```diff
@@ -33,4 +31,4 @@ execute: |
     done
     test -e ~/snap/test-snapd-xdg-autostart/current/foo-autostarted
 
-    test "$(get_journalctl_log_from_cursor "$cursor" | grep -c foo.desktop)" -gt 0
+    test "$("$TESTSTOOLS"/journal-state get-log_from_cursor "$cursor" | grep -c foo.desktop)" -gt 0
```
Comment (anonymouse64) [inline tests/main/snap-userd-desktop-app-autostart/task.yaml:34]: shouldn't this be 

```suggestion
    test "$("$TESTSTOOLS"/journal-state get-log-from-cursor "$cursor" | grep -c foo.desktop)" -gt 0
```
for consistency's sake ?

File: tests/lib/pkgdb.sh [hunk 14]
```diff
@@ -637,6 +637,7 @@ pkg_dependencies_ubuntu_classic(){
                 fwupd
                 packagekit
                 qemu-utils
+                shellcheck
```
Comment (anonymouse64) [inline tests/lib/pkgdb.sh:640]: why this change in this PR? is this related at all to the journalctl change? If not I would rather this be part of a different PR or otherwise called out in the PR description as a drive-by change.

Comment (anonymouse64) [reply tests/lib/pkgdb.sh:640]: ah is this needed for the test for journal-state itself? If so I would rather see us install shellcheck for just that test rather than for all systems as that will add time / network resources to every run unnecessarily

File: tests/lib/tools/suite/journal-state/task.yaml [hunk 15]
```diff
@@ -0,0 +1,37 @@
+summary: smoke test for the journal-state tool
+
+execute: |
+    # Run shellcheck for journal-state script
+    if [ "$SPREAD_SYSTEM" = "ubuntu-18.04-64" ]; then
+        shellcheck "$TESTSTOOLS"/journal-state
```
Comment (anonymouse64) [inline tests/lib/tools/suite/journal-state/task.yaml:6]: as mentioned above I think this test should install shellcheck manually here

File: tests/lib/tools/suite/journal-state/task.yaml [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+
+    # Check that using the last cursor the lines saved during prepare are not considered
+    cursor1=$("$TESTSTOOLS"/journal-state get-last-cursor)
+    test "$(journalctl --cursor "$cursor1" | grep -c "$SPREAD_JOB")" -eq 0
+
+    # Check the test is correctly started in the journal log
+    "$TESTSTOOLS"/journal-state start-new-log
+    retry --wait 1 --attempts 30 sh -c "test $(journalctl --cursor "$cursor1" | grep -c "$SPREAD_JOB") -eq 1"
+
+    # Check the subcommand check-log-started works well
```
Comment (anonymouse64) [inline tests/lib/tools/suite/journal-state/task.yaml:22]:
```suggestion
    # Check that the subcommand check-log-started works
```

File: tests/lib/tools/suite/journal-state/task.yaml [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+    # Check help
+    "$TESTSTOOLS"/journal-state | MATCH "usage: journal-state start-new-log"
+    "$TESTSTOOLS"/journal-state -h | MATCH "usage: journal-state start-new-log"
+    "$TESTSTOOLS"/journal-state --help | MATCH "usage: journal-state start-new-log"
+
+    # Check that using the last cursor the lines saved during prepare are not considered
+    cursor1=$("$TESTSTOOLS"/journal-state get-last-cursor)
+    test "$(journalctl --cursor "$cursor1" | grep -c "$SPREAD_JOB")" -eq 0
+
+    # Check the test is correctly started in the journal log
```
Comment (anonymouse64) [inline tests/lib/tools/suite/journal-state/task.yaml:18]:
```suggestion
    # Check that the test is correctly started in the journal log
```

File: tests/lib/tools/suite/journal-state/task.yaml [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+    "$TESTSTOOLS"/journal-state start-new-log
+    retry --wait 1 --attempts 30 sh -c "test $(journalctl --cursor "$cursor1" | grep -c "$SPREAD_JOB") -eq 1"
+
+    # Check the subcommand check-log-started works well
+    # The subcommand check-log-started could fail if the log does not contain the current SPREAD_JOB information
+    "$TESTSTOOLS"/journal-state check-log-started
+
+    # Check the subcommand get-log works well
```
Comment (anonymouse64) [inline tests/lib/tools/suite/journal-state/task.yaml:26]:
```suggestion
    # Check that the subcommand get-log works
```

File: tests/lib/tools/suite/journal-state/task.yaml [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+    echo "TEST-XX1" | systemd-cat -t snapd-test
+    retry --wait 1 --attempts 30 sh -c "test $("$TESTSTOOLS"/journal-state get-log | grep -c TEST-XX1) -eq 1"
+
+    # Check the subcommand get-last-cursor works
```
Comment (anonymouse64) [inline tests/lib/tools/suite/journal-state/task.yaml:30]:
```suggestion
    # Check that the subcommand get-last-cursor works
```

File: tests/lib/tools/suite/journal-state/task.yaml [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+    echo "TEST-XX2" | systemd-cat -t snapd-test
+    cursor2=$("$TESTSTOOLS"/journal-state get-last-cursor)
+    test "$("$TESTSTOOLS"/journal-state get-log-from-cursor "$cursor2" | grep -c "TEST-XX1")" -eq 0
+
+    # Check the subcommand match-log works well
```
Comment (anonymouse64) [inline tests/lib/tools/suite/journal-state/task.yaml:35]:
```suggestion
    # Check that the subcommand match-log works
```

File: tests/lib/tools/suite/journal-state/task.yaml [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+    "$TESTSTOOLS"/journal-state check-log-started
+
+    # Check the subcommand get-log works well
+    echo "TEST-XX1" | systemd-cat -t snapd-test
+    retry --wait 1 --attempts 30 sh -c "test $("$TESTSTOOLS"/journal-state get-log | grep -c TEST-XX1) -eq 1"
+
+    # Check the subcommand get-last-cursor works
+    echo "TEST-XX2" | systemd-cat -t snapd-test
+    cursor2=$("$TESTSTOOLS"/journal-state get-last-cursor)
+    test "$("$TESTSTOOLS"/journal-state get-log-from-cursor "$cursor2" | grep -c "TEST-XX1")" -eq 0
```
Comment (anonymouse64) [inline tests/lib/tools/suite/journal-state/task.yaml:33]: shouldn't this be checking for TEST-XX2 ?

```suggestion
    test "$("$TESTSTOOLS"/journal-state get-log-from-cursor "$cursor2" | grep -c "TEST-XX2")" -eq 0
```

File: tests/core/enable-disable-units-gpio/task.yaml (hunk 12 above)
Comment (sergiocazzolato) [reply tests/core/enable-disable-units-gpio/task.yaml:62]: I updated this because I found it weird, but shell check should complain for sure.
Also I'll add to my TODO to update the places where retry could be used.

File: tests/lib/tools/suite/journal-state/task.yaml (hunk 21 above)
Comment (sergiocazzolato) [reply tests/lib/tools/suite/journal-state/task.yaml:33]: No, I did that because I need something else in the log to make sure TEST-XX1 is not the last line, otherwise the cursor will point to TEST-XX1 and the grep -c will count 1 instead of 0

File: tests/lib/tools/suite/journal-state/task.yaml (hunk 18 above)
Comment (sergiocazzolato) [reply tests/lib/tools/suite/journal-state/task.yaml:26]: done

File: tests/lib/tools/suite/journal-state/task.yaml (hunk 19 above)
Comment (sergiocazzolato) [reply tests/lib/tools/suite/journal-state/task.yaml:30]: done

File: tests/lib/tools/suite/journal-state/task.yaml (hunk 20 above)
Comment (sergiocazzolato) [reply tests/lib/tools/suite/journal-state/task.yaml:35]: done

File: tests/lib/tools/suite/journal-state/task.yaml (hunk 17 above)
Comment (sergiocazzolato) [reply tests/lib/tools/suite/journal-state/task.yaml:18]: done

File: tests/lib/tools/suite/journal-state/task.yaml (hunk 16 above)
Comment (sergiocazzolato) [reply tests/lib/tools/suite/journal-state/task.yaml:22]: done

File: tests/lib/pkgdb.sh (hunk 14 above)
Comment (sergiocazzolato) [reply tests/lib/pkgdb.sh:640]: Well in this case it is going to be preinstalled in the image so we won't need to install it anymore. 
Once this PR is merged, then the spread-images project during the weekend will create a test image with all the dependencies for each system, so it is going to install in the image the shellcheck package. Does it make sense?

File: tests/lib/tools/suite/journal-state/task.yaml (hunk 15 above)
Comment (sergiocazzolato) [reply tests/lib/tools/suite/journal-state/task.yaml:6]: see previous answer

File: tests/main/snap-userd-desktop-app-autostart/task.yaml (hunk 13 above)
Comment (sergiocazzolato) [reply tests/main/snap-userd-desktop-app-autostart/task.yaml:34]: Nice catch, thans for the review!!

File: tests/lib/tools/suite/journal-state/task.yaml (hunk 21 above)
Comment (anonymouse64) [reply tests/lib/tools/suite/journal-state/task.yaml:33]: ah I see what you mean, this is fine then

File: tests/lib/pkgdb.sh (hunk 14 above)
Comment (anonymouse64) [reply tests/lib/pkgdb.sh:640]: ah if it's preinstalled in the image that's fine then

File: tests/core/enable-disable-units-gpio/task.yaml (hunk 12 above)
Comment (anonymouse64) [reply tests/core/enable-disable-units-gpio/task.yaml:62]: yeah this isn't a big deal, either way you want to handle it is fine

File: tests/lib/tools/suite/journal-state/task.yaml (hunk 1 above)
Comment (zyga) [reply tests/lib/tools/suite/journal-state/task.yaml:4]: Where is this checked?

**Comment by sergiocazzolato** [comment]
@zyga, the shellcheck verification for all the tools is done in unit/go test just on ubuntu-18.04-64

File: tests/lib/tools/suite/journal-state/task.yaml (hunk 1 above)
Comment (sergiocazzolato) [reply tests/lib/tools/suite/journal-state/task.yaml:4]: Sorry, it was included in another pr, now the we run static cehcks through run-checks script all the tools are validated.
Currently the static checks are validated when running the unit/go spread test. Shellcheck package is preinstalled for ubuntu-18.04-64 so we make sure the validation is done.

File: tests/lib/tools/journal-state [hunk 22]
```diff
@@ -1,19 +1,25 @@
...
-JOURNALCTL_CURSOR_FILE="$RUNTIME_STATE_PATH"/journalctl_cursor
+JOURNALCTL_CURSOR_FILE="$SPREAD_PATH"/tests/runtime-state/journalctl_cursor
+
+show_help() {
+    echo "usage: journal-state start-new-log"
+    echo "usage: journal-state check-log-started"
+    echo "usage: journal-state get-last-cursor"
+    echo "usage: journal-state match-log [expression]"
+    echo "usage: journal-state get-log"
+    echo "usage: journal-state get-log-from-cursor"
```
Comment (bboozzoo) [inline tests/lib/tools/journal-state:12]:
```suggestion
    echo "usage: journal-state start-new-log"
    echo "       journal-state check-log-started"
    echo "       journal-state get-last-cursor"
    echo "       journal-state match-log [expression]"
    echo "       journal-state get-log"
    echo "       journal-state get-log-from-cursor"
```

File: tests/lib/tools/suite/journal-state/task.yaml (hunk 1 above)
Comment (bboozzoo) [reply tests/lib/tools/suite/journal-state/task.yaml:4]: FWIW, journal-state is shellcheck clean atm.

File: tests/lib/tools/journal-state (hunk 22 above)
Comment (bboozzoo) [reply tests/lib/tools/journal-state:12]: I'll push this change.

**Comment by zyga** [comment]
The pre-seed failure in 20.10 looks suspicious. I'll check what's going on there.

**Comment by zyga** [comment]
The preseed reset test is also failing on master. I'm inclined to merge this.

**Comment by anonymouse64** [comment]
> The pre-seed failure in 20.10 looks suspicious. I'll check what's going on there.

the preseed failures are known and something we're working on this week, but haven't resolved quite yet, but the gist of it is that the groovy images are now preseeded, so our preseed tests can't really preseed them anymore.
